rust/protocol/src/sealed_sender.rs

```diff
@@ -0,0 +1,784 @@
+        let pb = proto::sealed_sender::UnidentifiedSenderMessage {
+            ephemeral_public: Some(ephemeral_public.serialize().to_vec()),
+            encrypted_static: Some(encrypted_static.clone()),
+            encrypted_message: Some(encrypted_message.clone()),
+        };
+        pb.encode(&mut serialized)?; // appends to buffer
+
+        Ok(Self {
+            version,
+            ephemeral_public,
+            encrypted_static,
+            encrypted_message,
+            serialized,
+        })
+    }
+
+    pub fn version(&self) -> Result<u8> {
+        Ok(self.version)
+    }
+
+    pub fn ephemeral_public(&self) -> Result<PublicKey> {
+        Ok(self.ephemeral_public)
+    }
+
+    pub fn encrypted_static(&self) -> Result<&[u8]> {
+        Ok(&self.encrypted_static)
+    }
+
+    pub fn encrypted_message(&self) -> Result<&[u8]> {
+        Ok(&self.encrypted_message)
+    }
+
+    pub fn serialized(&self) -> Result<&[u8]> {
+        Ok(&self.serialized)
+    }
+}
+
+struct EphemeralKeys {
+    derived_values: Box<[u8]>,
+}
+
+impl EphemeralKeys {
+    pub fn calculate(
+        their_public: &PublicKey,
+        our_public: &PublicKey,
+        our_private: &PrivateKey,
+        sending: bool,
+    ) -> Result<Self> {
+        let mut ephemeral_salt = Vec::with_capacity(2 * 32 + 20);
+        ephemeral_salt.extend_from_slice("UnidentifiedDelivery".as_bytes());
+
+        if sending {
+            ephemeral_salt.extend_from_slice(&their_public.serialize());
+        }
+        ephemeral_salt.extend_from_slice(&our_public.serialize());
+        if !sending {
+            ephemeral_salt.extend_from_slice(&their_public.serialize());
+        }
+
+        let shared_secret = our_private.calculate_agreement(their_public)?;
+        let kdf = HKDF::new(3)?;
+        let derived_values = kdf.derive_salted_secrets(&shared_secret, &ephemeral_salt, &[], 96)?;
+
+        Ok(Self { derived_values })
+    }
+
+    fn chain_key(&self) -> Result<&[u8]> {
+        Ok(&self.derived_values[0..32])
+    }
+
+    fn cipher_key(&self) -> Result<&[u8]> {
+        Ok(&self.derived_values[32..64])
+    }
+
+    fn mac_key(&self) -> Result<&[u8]> {
+        Ok(&self.derived_values[64..96])
+    }
+}
+
+struct StaticKeys {
+    derived_values: Box<[u8]>,
+}
+impl StaticKeys {
+    pub fn calculate(
+        their_public: &PublicKey,
+        our_private: &PrivateKey,
+        chain_key: &[u8],
+        ctext: &[u8],
+    ) -> Result<Self> {
+        let mut salt = Vec::with_capacity(chain_key.len() + ctext.len());
+        salt.extend_from_slice(chain_key);
+        salt.extend_from_slice(ctext);
+
+        let shared_secret = our_private.calculate_agreement(their_public)?;
+        let kdf = HKDF::new(3)?;
+        let derived_values = kdf.derive_salted_secrets(&shared_secret, &salt, &[], 96)?;
+
+        Ok(Self { derived_values })
+    }
+
+    fn cipher_key(&self) -> Result<&[u8]> {
+        Ok(&self.derived_values[32..64])
```

Comment:
The first 32 bytes are just thrown away?

Reply:
Yes, which is odd and doesn't have any positive effect that I can see but that's how it works. I'll leave a comment to make it explicit.

(I considered throwing away the first 32 bytes in `new`, which would save a little memory at the cost of a second allocation, didn't seem worth it)